File: Python/RadixSort.py
def radixSort(arr):
    max_num = max(arr):
    exp = 1

    while max_num // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for i in range(n):
        index = arr[i] // exp
        count[index%10] += 1

    for i in range(1, 10):
        count [i] += count[i-1]

    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index%10]-1] = arr[i]
        count[index%10] -= 1
        i -= 1

    for i in range(n):
        arr[i] = output[i]

list = [64, 34, 25, 12, 22, 11, 90]
print(radixSort(list))

File: Cpp/RadixSort.cpp
#include <iostream>
#include <vector>
#include "RadixSort.h"

using namespace std;

int getMax(vector<int>& arr) {
    int max = arr[0];
    for (int i = 1; i < arr.size(); ++i) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    return max;
}

void countingSort(vector<int>& arr, int exp) {
    const int radix = 10; // Base 10 (decimal system)
    int n = arr.size();
    vector<int> output(n);
    vector<int> count(radix, 0);

    // Count occurrences of digits in the input array
    for (int i = 0; i < n; ++i) {
        int digit = (arr[i] / exp) % radix;
        count[digit]++;
    }

    // Change count[i] so that count[i] contains the actual
    // position of this digit in output[]
    for (int i = 1; i < radix; ++i) {
        count[i] += count[i - 1];
    }

    // Build the output array
    for (int i = n - 1; i >= 0; --i) {
        int digit = (arr[i] / exp) % radix;
        output[count[digit] - 1] = arr[i];
        count[digit]--;
    }

    // Copy the output array to the input array so that arr[] now contains sorted numbers
    for (int i = 0; i < n; ++i) {
        arr[i] = output[i];
    }
}

void radixSort(vector<int>& arr) {
    int max = getMax(arr);

    // Do counting sort for every digit. Note that
    // instead of passing digit number, exp is passed.
    // exp is 10^i where i is the current digit number
    for (int exp = 1; max / exp > 0; exp *= 10) {
        countingSort(arr, exp);
    }
}